lib/rules/sort-alternatives.ts

```diff
@@ -0,0 +1,396 @@
+                ranges.push(...context.toCharSet(d).ranges)
+            } else if (d.type === "Backreference" && !isEmptyBackreference(d)) {
+                ranges.push(...getConsumedChars(d.resolved, context).ranges)
+            }
+
+            // always continue to the next element
+            return false
+        },
+        // don't go into assertions
+        (d) => d.type !== "Assertion" && d.type !== "CharacterClass",
+    )
+
+    return Chars.empty(context.flags).union(ranges)
+}
+
+type Parent = Group | CapturingGroup | Pattern | LookaroundAssertion
+
+/**
+ * Assuming that the given group only consumes the given characters, this will
+ * return whether the alternatives of the group can be reordered freely without
+ * affecting the behavior of the regex.
+ *
+ * This also assumes that the alternatives of the given group do not contain
+ * capturing group in such a way that their order matters.
+ */
+function canReorder(
+    parent: Parent,
+    consumedChars: CharSet,
+    context: RegExpContext,
+): boolean {
+    const lengthRange = getLengthRange(parent.alternatives)
+    if (lengthRange && lengthRange.min === lengthRange.max) {
+        return true
+    }
+
+    if (parent.type === "Pattern" || parent.type === "Assertion") {
+        return false
+    }
+
+    return (
+        getFirstCharAfter(parent, "rtl", context.flags).char.isDisjointWith(
+            consumedChars,
+        ) &&
+        getFirstCharAfter(parent, "ltr", context.flags).char.isDisjointWith(
+            consumedChars,
+        )
+    )
+}
+
+/**
+ * Returns whether the given element contains only literal characters and
+ * groups/other elements containing literal characters.
+ */
+function containsOnlyLiterals(
+    element: Element | Pattern | Alternative,
+): boolean {
+    return !hasSomeDescendant(
+        element,
+        (d) => {
+            return (
+                d.type === "Backreference" ||
+                d.type === "CharacterSet" ||
+                (d.type === "Quantifier" && d.max === Infinity) ||
+                (d.type === "CharacterClass" && d.negate)
+            )
+        },
+        (d) => d.type !== "Assertion",
+    )
+}
+
+/**
+ * Sorts the given alternatives.
+ */
+function sortAlternatives(
+    alternatives: Alternative[],
+    context: RegExpContext,
+): void {
+    const firstChars = new Map<Alternative, number>()
+    for (const a of alternatives) {
+        const chars = getFirstConsumedChar(a, "ltr", context.flags)
+        const char =
+            chars.empty || chars.char.isEmpty
+                ? Infinity
+                : chars.char.ranges[0].min
+        firstChars.set(a, char)
+    }
+
+    alternatives.sort((a, b) => {
+        const firstA = firstChars.get(a)!
+        const firstB = firstChars.get(b)!
+        if (firstA !== firstB) {
+            return firstA - firstB
+        }
+
+        if (context.flags.ignoreCase) {
+            return a.raw.toUpperCase().localeCompare(b.raw.toUpperCase())
```

Comment:
What do you think about making the following change?

```suggestion
            return (
                a.raw.toUpperCase().localeCompare(b.raw.toUpperCase()) ||
                a.raw.localeCompare(b.raw)
            )
```

With this change, the following two patterns enforce a consistent order:

`/\b(?:A|a|C|c|B|b)\b/i` -> `/\b(?:a|A|b|B|c|C)\b/i`
`/\b(?:a|A|C|c|b|B)\b/i` -> `/\b(?:a|A|b|B|c|C)\b/i`

Before the change, it returned the following results:

`/\b(?:A|a|C|c|B|b)\b/i` -> `/\b(?:A|a|B|b|C|c)\b/i`
`/\b(?:a|A|C|c|b|B)\b/i` -> `/\b(?:a|A|b|B|C|c)\b/i`

However, after the change, the following pattern will result in an error. This may be annoying to the user.

`/\b(?:A|a|B|b|C|c)\b/i`

Reply:
Very good point. You also just pointed out a bug. 

Right now, alternatives are sorted by the first character first and then by raw. But the first character is ordered by byte order (by ascending Unicode code point value; "A" < "a"), and the raw is sorted by the current locale ("a" < "A").

I will use your suggestion and implement a byte-order compare function.